agent/src/com/jonnyzzz/teamcity/virtual/run/docker/DockerVM.java

```diff
@@ -125,12 +139,14 @@ public BuildProcess startImpl() throws RunBuildException {
       }
 
       @NotNull
-      private List<String> scriptRun(@NotNull final File script) {
-        return Arrays.asList(
-                ctx.getShellLocation(),
-                "-c",
-                "\"source " + script.getName() + "\""
-        );
+      private List<String> scriptRun(@NotNull final File script) throws RunBuildException {
+        if(ctx.isDockerServerWindowsBased() && ctx.getShellLocationInsideContainer().equals("cmd.exe")) {
+          return  Arrays.asList(ctx.getShellLocationInsideContainer(), "/C", "call " + script.getName());
+        } else if(ctx.isDockerServerWindowsBased() && ctx.getShellLocationInsideContainer().equals("powershell.exe")) {
+          return  Arrays.asList(ctx.getShellLocationInsideContainer(), "-Command", "get-content " + script.getName() + " | Invoke-Expression");
```

Comment:
As far as I remember, it's more stable to call script as a file instead of doing `Invoke-Expression`. The problem was with missing exit code as in script mode it was not failing asap. Could you please check issues for powershell runner in TeamCity tracker for details. This should be an opensource plugin, btw.